### PR TITLE
Session api

### PR DESCRIPTION
**Pull request by sravfeyn**
This removes session based access to our public API https://trello.com/c/0xqVXNJX/581-security-remove-session-based-api-access
@orangejenny 
FYI @millerdev @czue

### CODE REVIEW

File: corehq/apps/domain/decorators.py [hunk 1]
```diff
@@ -161,22 +161,26 @@ def _inner(request, domain, *args, **kwargs):
...
+digest_auth = login_or_digest_ex(allow_sessions=False)
 
 
-def login_or_basic_ex(allow_cc_users=False):
-    return _login_or_challenge(basicauth(), allow_cc_users=allow_cc_users)
+def login_or_basic_ex(allow_cc_users=False, allow_sessions=True):
+    return _login_or_challenge(basicauth(), allow_cc_users=allow_cc_users, allow_sessions=allow_sessions)
 
 login_or_basic = login_or_basic_ex()
+basic_auth =  login_or_basic_ex(allow_sessions=False)
```
Comment (dimagimon) [inline corehq/apps/domain/decorators.py]: E222 multiple spaces after operator

File: corehq/apps/api/tests.py [hunk 2]
```diff
@@ -125,6 +126,34 @@ def single_endpoint(self, id):
...
+        response = self.client.get(session_url)
+        self.assertEqual(response.status_code, 403)
+
+        # api_key auth should succeed
+        url_params.update({'username': username, 'api_key': api_key})
+        api_url = "%s?%s" % (url, urlencode(url_params))
+        print api_url, "binbong"
+        response = self.client.get(api_url)
+        if response.status_code != 200:
+            import pdb; pdb.set_trace()
```
Comment (dimagimon) [inline corehq/apps/api/tests.py]: E702 multiple statements on one line (semicolon)

**Comment by dimagimon** [comment]
There are 13 errors:

* corehq/apps/api/tests.py, line 236 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 302 - E303 too many blank lines (2)
E261 at least two spaces before inline comment
E501 line too long (116 > 115 characters)
* corehq/apps/api/tests.py, line 840 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 204 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 1239 - E501 line too long (126 > 115 characters)
* corehq/apps/api/tests.py, line 610 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 1246 - E501 line too long (144 > 115 characters)
* corehq/apps/api/tests.py, line 372 - E303 too many blank lines (3)
* corehq/apps/api/tests.py, line 306 - E261 at least two spaces before inline comment
E501 line too long (117 > 115 characters)
* corehq/apps/api/tests.py, line 1365 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 736 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 1274 - E501 line too long (130 > 115 characters)
* corehq/apps/api/tests.py, line 1266 - E501 line too long (127 > 115 characters)

**Comment by dimagimon** [comment]
There are 13 errors:

* corehq/apps/api/tests.py, line 205 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 1366 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 611 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 237 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 841 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 1240 - E501 line too long (126 > 115 characters)
* corehq/apps/api/tests.py, line 737 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 373 - E303 too many blank lines (3)
* corehq/apps/api/tests.py, line 1267 - E501 line too long (127 > 115 characters)
* corehq/apps/api/tests.py, line 307 - E261 at least two spaces before inline comment
E501 line too long (117 > 115 characters)
* corehq/apps/api/tests.py, line 1275 - E501 line too long (130 > 115 characters)
* corehq/apps/api/tests.py, line 303 - E303 too many blank lines (2)
E261 at least two spaces before inline comment
E501 line too long (116 > 115 characters)
* corehq/apps/api/tests.py, line 1247 - E501 line too long (144 > 115 characters)

**Comment by dimagimon** [comment]
There are 13 errors:

* corehq/apps/api/tests.py, line 205 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 1366 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 611 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 237 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 841 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 1240 - E501 line too long (126 > 115 characters)
* corehq/apps/api/tests.py, line 737 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 373 - E303 too many blank lines (3)
* corehq/apps/api/tests.py, line 1267 - E501 line too long (127 > 115 characters)
* corehq/apps/api/tests.py, line 307 - E261 at least two spaces before inline comment
E501 line too long (117 > 115 characters)
* corehq/apps/api/tests.py, line 1275 - E501 line too long (130 > 115 characters)
* corehq/apps/api/tests.py, line 303 - E303 too many blank lines (2)
E261 at least two spaces before inline comment
E501 line too long (116 > 115 characters)
* corehq/apps/api/tests.py, line 1247 - E501 line too long (144 > 115 characters)

File: corehq/apps/domain/decorators.py [hunk 3]
```diff
@@ -161,22 +166,16 @@ def _inner(request, domain, *args, **kwargs):
                         return HttpResponseForbidden()
 
                 return _inner(request, domain, *args, **kwargs)
-            else:
+            elif allow_sessions:
```
Comment (sravfeyn) [inline corehq/apps/domain/decorators.py]: The main change of the PR

**Comment by millerdev** [comment]
test failures look real:
```
ERROR: test_basic (corehq.apps.api.tests.TestBulkUserAPI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/api/tests.py", line 1491, in test_basic
    response = self.query()
  File "/mnt/corehq/apps/api/tests.py", line 1476, in query
    return self._assert_auth_get_resource(url)
  File "/mnt/corehq/apps/api/tests.py", line 158, in _assert_auth_get_resource
    api_url = self._api_url(url, username)
  File "/mnt/corehq/apps/api/tests.py", line 134, in _api_url
    api_key = self.api_key.key
AttributeError: 'TestBulkUserAPI' object has no attribute 'api_key'
```

```
FAIL: If correct credentials for a commcare user are submitted, the response is the record for that commcare user
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/api/tests.py", line 1247, in test_commcare_user_success
    response = self._assert_auth_post_resource(self.list_endpoint, {'username': self.commcare_username, 'password': self.commcare_password})
  File "/mnt/corehq/apps/api/tests.py", line 169, in _assert_auth_post_resource
    self.assertEqual(response.status_code, 401)
AssertionError: 400 != 401
```

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/apps/api/tests.py, line 205 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 611 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 237 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 841 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 737 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 373 - E303 too many blank lines (3)
* corehq/apps/api/tests.py, line 307 - E261 at least two spaces before inline comment
E501 line too long (117 > 115 characters)
* corehq/apps/api/tests.py, line 1368 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 303 - E303 too many blank lines (2)
E261 at least two spaces before inline comment
E501 line too long (116 > 115 characters)

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/apps/api/tests.py, line 205 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 611 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 237 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 841 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 737 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 373 - E303 too many blank lines (3)
* corehq/apps/api/tests.py, line 307 - E261 at least two spaces before inline comment
E501 line too long (117 > 115 characters)
* corehq/apps/api/tests.py, line 1368 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 303 - E303 too many blank lines (2)
E261 at least two spaces before inline comment
E501 line too long (116 > 115 characters)

File: corehq/apps/api/tests.py [hunk 4]
```diff
@@ -125,6 +126,56 @@ def single_endpoint(self, id):
                                                           resource_name=self.resource.Meta.resource_name,
                                                           pk=id))
 
+    def _api_url(self, url, username=None):
+        if 'api_key' in url:
```
Comment (orangejenny) [inline corehq/apps/api/tests.py]: This could be a bit more robust (parse the url and check that `api_key` is a param), but "api_key" isn't that likely to show up elsewhere in the url, so up to you if you want to change it.

Comment (orangejenny) [inline corehq/apps/api/tests.py]: oh, nm, this is just for tests

File: corehq/apps/api/tests.py [hunk 5]
```diff
@@ -125,6 +126,56 @@ def single_endpoint(self, id):
...
+        self.client.login(username=username, password=password)
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 401)
+
+        # api_key auth should succeed, caller can check for expected code
+        api_url = self._api_url(url, username)
+        response = self.client.get(api_url)
+        return response
+
+    def _assert_auth_post_resource(self, url, post_data, content_type='application/json', method="POST"):
```
Comment (orangejenny) [inline corehq/apps/api/tests.py]: nit, but splitting this into `_assert_auth_post_resource` and `_assert_auth_put_resource` would be about as clean as taking a method param (which is a little weird when "post" is in the same of the method)

**Comment by orangejenny** [comment]
Looks good to me, kicked build.

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/apps/api/tests.py, line 205 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 237 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 737 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 1368 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 373 - E303 too many blank lines (3)
* corehq/apps/api/tests.py, line 307 - E261 at least two spaces before inline comment
E501 line too long (117 > 115 characters)
* corehq/apps/api/tests.py, line 611 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 303 - E303 too many blank lines (2)
E261 at least two spaces before inline comment
E501 line too long (116 > 115 characters)
* corehq/apps/api/tests.py, line 841 - E126 continuation line over-indented for hanging indent

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/apps/api/tests.py, line 205 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 737 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 237 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 373 - E303 too many blank lines (3)
* corehq/apps/api/tests.py, line 307 - E261 at least two spaces before inline comment
E501 line too long (117 > 115 characters)
* corehq/apps/api/tests.py, line 841 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 611 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 303 - E303 too many blank lines (2)
E261 at least two spaces before inline comment
E501 line too long (116 > 115 characters)
* corehq/apps/api/tests.py, line 1368 - E126 continuation line over-indented for hanging indent

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/apps/api/tests.py, line 205 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 737 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 237 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 373 - E303 too many blank lines (3)
* corehq/apps/api/tests.py, line 307 - E261 at least two spaces before inline comment
E501 line too long (117 > 115 characters)
* corehq/apps/api/tests.py, line 841 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 611 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 303 - E303 too many blank lines (2)
E261 at least two spaces before inline comment
E501 line too long (116 > 115 characters)
* corehq/apps/api/tests.py, line 1368 - E126 continuation line over-indented for hanging indent

File: corehq/apps/api/resources/v0_1.py [hunk 6]
```diff
@@ -270,7 +270,7 @@ class Meta(UserResource.Meta):
 
     def obj_get_list(self, bundle, **kwargs):
         domain = kwargs['domain']
-        username = bundle.request.GET.get('username')
+        username = bundle.request.GET.get('web_username')
```
Comment (sravfeyn) [inline corehq/apps/api/resources/v0_1.py]: It's not great to change API behaviour midway, but here is the reasoning.

`username` GET param is used by tastypie authentication. So using 'username' param for our use is resulting in unexpected behaviour (for e.g. webusers get filtered by `username`, when `username` is passed for authentication purpose). So I am renaming this to `web_username`. We might break existing use cases, but for following reason I think it's okay
1. As we discussed at the start of this PR, we assumed that users are expected to use api key auth only, so this conflict has been there for a while and no one reported an error.

@czue @millerdev

Comment (czue) [inline corehq/apps/api/resources/v0_1.py]: what happens if you currently use basic or digest auth with a username param? if that used to work then i think we should at least check the proxy logs and see if anyone's doing that. should be quite easy to see and if it is in use then notify people who are using it

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/apps/api/tests.py, line 205 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 737 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 237 - E303 too many blank lines (2)
* corehq/apps/api/tests.py, line 373 - E303 too many blank lines (3)
* corehq/apps/api/tests.py, line 307 - E261 at least two spaces before inline comment
E501 line too long (117 > 115 characters)
* corehq/apps/api/tests.py, line 841 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 611 - E126 continuation line over-indented for hanging indent
* corehq/apps/api/tests.py, line 303 - E303 too many blank lines (2)
E261 at least two spaces before inline comment
E501 line too long (116 > 115 characters)
* corehq/apps/api/tests.py, line 1368 - E126 continuation line over-indented for hanging indent

File: corehq/apps/api/resources/v0_1.py (hunk 6 above)
Comment (millerdev) [inline corehq/apps/api/resources/v0_1.py]: Why does this need to change in this PR? Is it necessary for other things that are happening in this PR or is it fixing an unrelated bug? If unrelated then it might make sense to do in a separate PR so it can be easily rolled back if it turns out to be a problem.

I would not expect this to affect basic or digest auth since the credentials for those are not in GET parameters.

Comment (sravfeyn) [inline corehq/apps/api/resources/v0_1.py]: @czue Yeah it doesn't affect basic or digest. I checked on proxy logs, there are no real requests to `web-user` endpoint apart from (me and ethan for testing). The log file is since 13th May. Should I be checking even older logs?

@millerdev The only reason I am doing in this PR is because the bug is uncovered due to tests written in this PR. The bug is related to this PR

Comment (czue) [inline corehq/apps/api/resources/v0_1.py]: if this doesn't affect anyone who used the API since May 13 then I think it's fine from my side

File: corehq/apps/api/tests.py [hunk 7]
```diff
@@ -801,9 +832,10 @@ def test_update(self):
                 "type": cls.__name__,
                 "url": "http://example.com/forwarding/modified/{cls}".format(cls=cls.__name__),
             }
-            response = self.client.put(self.single_endpoint(backend_id),
+            response = self._assert_auth_post_resource(self.single_endpoint(backend_id),
                                        json.dumps(repeater_json),
-                                       content_type='application/json')
+                                       content_type='application/json',
```
Comment (dimagimon) [inline corehq/apps/api/tests.py]: E126 continuation line over-indented for hanging indent

**Comment by sravfeyn** [comment]
This is good to go if no other concerns

File: corehq/apps/api/resources/v0_1.py (hunk 6 above)
Comment (millerdev) [inline corehq/apps/api/resources/v0_1.py]: 👍

File: corehq/apps/domain/decorators.py [hunk 8]
```diff
@@ -161,22 +166,16 @@ def _inner(request, domain, *args, **kwargs):
...
                 return login_and_domain_required(fn)(request, domain, *args, **kwargs)
+            else:
+                return HttpResponseForbidden()
```
Comment (millerdev) [inline corehq/apps/domain/decorators.py]: I think this else clause will never be executed. It also seems easier to understand if the first condition is inverted (stated positively). Here's a series of transformations that I believe are logically consistent and end up clearer:

```py
if not request.user.is_authenticated() or not allow_sessions:
    ... # make _innter that checks stuff, maybe calls fn
    return _inner(request, domain, *args, **kwargs)
elif allow_sessions:
    return login_and_domain_required(fn)(request, domain, *args, **kwargs)
else:
    return HttpResponseForbidden()
```
is the same as (`elif` -> `else` with nested `if`)
```py
if not request.user.is_authenticated() or not allow_sessions:
    ... # make _innter that checks stuff, maybe calls fn
    return _inner(request, domain, *args, **kwargs)
else:
    if allow_sessions:
        return login_and_domain_required(fn)(request, domain, *args, **kwargs)
    else:
        return HttpResponseForbidden()
```
is the same as (invert first if condition)
```py
if request.user.is_authenticated() and allow_sessions:
    if allow_sessions:
        return login_and_domain_required(fn)(request, domain, *args, **kwargs)
    else:
        return HttpResponseForbidden()
else:
    ... # make _innter that checks stuff, maybe calls fn
    return _inner(request, domain, *args, **kwargs)
```
is the same as (drop redundant condition)
```py
if request.user.is_authenticated() and allow_sessions:
    return login_and_domain_required(fn)(request, domain, *args, **kwargs)
else:
    ... # make _innter that checks stuff, maybe calls fn
    return _inner(request, domain, *args, **kwargs)
```

**Comment by sravfeyn** [comment]
@millerdev Updated

**Comment by czue** [comment]
this causes an authentication box to pop up anytime you load anything that uses the drilldown location filter (commtrack reports). I'm looking into a quick fix but might be worth reverting if I don't find one soon

**Comment by czue** [comment]
the fix seems simple enough (replace the new auth methods with the old ones for the `location_internal` API, but I'm now concerned there wasn't a thorough enough audit of the internal API usage before this went out so thinking it is still worth reverting

**Comment by sravfeyn** [comment]
Looking into this
